controlpanel/api/slack.py

```diff
@@ -0,0 +1,20 @@
+from django.conf import settings
+import slack
+
+
+CREATE_SUPERUSER_MESSAGE = "`{username}` was created as a superuser"
+GRANT_SUPERUSER_ACCESS_MESSAGE = "`{username}` was granted superuser status"
```

Comment:
I think the "constants" are overkill, if we really want to have these as "constants" I'd prefer to have them close to the place where they're used as at the end of the day this `slack` module is not concerned with the specific messages sent, it just needs a string to sent to the team channel.

Reply:
I disagree. This module is the interface between the Control Panel and the Slack client. The Slack client is only concerned with sending strings to the channel. The rest of the Control Panel code is only concerned that a notification is sent. This module is concerned with how that notification is sent (via Slack) and how it is formatted.
The strings should be constants, because we test that they are sent and if we change the contents of the message (but not the reason they are sent), then we would also have to update the tests.
The constants belong in this module because they are Slack notifications, and not used for logging or flash messages, for example. If we add more Slack notifications on other events and kept the message strings alongside their use, we would have notification strings spread throughout the codebase instead of gathered here in the relevant Slack module.

Reply:
I think we disagree then.

> This module is concerned with how that notification is sent (via Slack) and how it is formatted

Not really because I can `import slack; slack.notify_team("I like icecream")` and that's perfectly legit from the slack module perspective.

If you want this module to be more specific and have more control over the kind of messages sent to slack this should pull more weight, e.g.

```python
def notify_made_superuser(granted_by, username):
    # generate "{granted_by} made {username} a superuser"
    # use constants or whatever
    msg = "..."
    _notify_team(msg)

def _notify_team(msg):
    # just the logic to send messages to Slack
    pass
```

If we go this route (which may have sense if you think the responsibility of this module is to control which notifications are sent to Slack) this means the logic to add the `by John Doe` would go in here rather than in the model and I'd like that.

Reply:
Before I forget, is it actually possible for a superuser to create another superuser before they even logged in? I thought a user was created only when they logged in the first time (sorry if I'm missing something), if that's the case you may not need the `CREATE_SUPERUSER_MESSAGE` and you could possibly simplify the logic a bit.

Reply:
You are right about having a function like `notify_superuser_created(username, by_username)` - I will do that.
It is possible to create a new user as a superuser before they login programmatically or via the shell - this would also create their k8s namespace and IAM role, etc. You would have to know their Github ID.